x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -321,6 +323,40 @@ protected SpatialRelatesFunction rule(SpatialRelatesFunction function) {
         }
     }
 
+    static class ReplaceOrderByExpressionWithEval extends OptimizerRules.OptimizerRule<OrderBy> {
+
+        @Override
+        protected LogicalPlan rule(OrderBy orderBy) {
+            int counter = 0;
+            List<Alias> evals = new ArrayList<>(orderBy.order().size());
+            List<Order> newOrders = new ArrayList<>(orderBy.order().size());
+
+            for (Order order : orderBy.order()) {
+                if (order.child() instanceof Attribute == false) {
+                    var name = rawTemporaryName("order_by", "temp_name", String.valueOf(counter++));
```

Comment:
Maybe find a better name than "temp_name" - such as the ordinal of the order it tries to replace?